tests/enrol_advanced_testcase_test.php

```diff
@@ -0,0 +1,330 @@
+     $evento_personen_anmeldung = $builder->add_personen_anmeldung("2019-02-17T00:00:00.000+01:00", "hoferlis", "2018-06-05T08:58:20.723+02:00","auto" ,415864, 20216, 25491, 118000, $evento_status);
+     $evento_person = $builder->add_person("Teacher", "Mister", "mister.teacher@stud.htwchur.ch",  118000, 30040, true, 118000, $evento_personen_anmeldung);
+     $ad_account = $builder->add_ad_account(0, "2019-02-17T00:00:00.000+01:00", "2019-02-17T00:00:00.000+01:00", 0, 118000, 0, 1, 0, "S-1-5-21-2460181394-1097805571-3701207438-51315", "MisTe");
+
+
+     $this->simulator =$builder->service;
+     $this->locallib = new enrol_evento_user_sync_exposed($this->simulator);
+     $this->resetAfterTest(true);
+     $this->create_moodle_course();
+   }
+
+   /*Enable plugin method*/
+   protected function enable_plugin(){
+     $enabled = enrol_get_plugins(true);
+     $enabled['evento'] = true;
+     $enabled = array_keys($enabled);
+     set_config('enrol_plugins_enabled', implode(',', $enabled));
+   }
+   /*disable plugin method*/
+   protected function disable_plugin(){
+     $enabled = enrol_get_plugins(true);
+     unset($enabled['evento']);
+     $enabled = array_keys($enabled);
+     set_config('enrol_plugins_enabled', implode(',', $enabled));
+   }
+   /*get enroled user from course*/
+   protected function get_enroled_user($id){
+     global $DB;
+     $this->user_enrolment = $DB->get_record('enrol', array('courseid'=>$id, 'enrol'=>'evento'), '*', MUST_EXIST);
+     $this->enrolments = $DB->count_records('user_enrolments', array('enrolid'=>$this->user_enrolment->id));
+  //   var_dump($this->user_enrolment->id);
+   }
+
+   /*Simuation test if plugin is enabled*/
+   /**
+   * @test
+   */
+   public function basic(){
+     $anlass = $this->simulator->get_event_by_number("mod.mmpAUKATE1.HS18_BS.002");
+     $personen_anmeldung = $this->simulator->get_enrolments_by_eventid(25490);
+     //var_dump($personen_anmeldung);
+     $personen_anmeldung = $this->simulator->get_enrolments_by_eventid(25490);
+
+     //var_dump($this->simulator->evento_personen_anmeldungen);
+
+     $person = $this->simulator->get_person_by_id(141703);
+     //var_dump($person);
+     $person = $this->simulator->get_person_by_id(117828);
+     //var_dump($person);
+     $ad_account = $this->simulator->get_ad_accounts_by_evento_personid(141701, null, null);
+     //var_dump($ad_account);
+     $ad_account_student = $this->simulator->get_all_ad_accounts(null);
+     //var_dump($ad_account_student);
+}
+
+   /*Basic test if plugin is enabled*/
+   /**
+   * @test
+   */
+   public function basics(){
+
+     $this->resetAfterTest(true);
+
+     $this->enable_plugin();
+     $plugin = 'evento';
+     $evento_plugin = enrol_get_plugin($plugin);
+
+     $this->assertEquals( $evento_plugin->get_name(), 'evento');
+     $this->assertNotEmpty( $evento_plugin);
+   }
+   /*get_user() Test for a new user*/
+   /**
+   * @test
+   */
+   public function get_user(){
```

Comment:
Jetzt habe ich mir gerade nochmals die Funktion get_user (https://github.com/thomasianwieling/moodle-enrol_evento/blob/master/locallib.php#L389), die ja zu testen ist, angeschaut.

Wenn ich den Code korrekt verstehe, hat diese Funktion mehrere Aufgaben
1. Prüfen, ob genau ein Benutzer mit diesem Account in der AD ist (Zeilen 400 - 412)
2. Dann den dazugehörigen Moodle-Benutzer zurückgeben
   
Der Schritt 2 hat noch Unterschritte:
1. Falls ein Moodle-User mit der EventoId existiert, dnan diesen zurückgeben.
2. Falls kein Moodle-User mit der EventoId existiert, dann einen neuen Moodle-Benutzer erstellen zu diesen zurückgeben.

Das sind also mindestens 3 Fälle, die man in testen sollte.
Aktuell

Reply:
Dann sehe hat es im Code auch noch Fehlerfälle, z.B. cannotfindaduser. Das müsste wiederzum einen Test geben:

Reply:
Genau in dieser Art und Weise müsste man die Tests umschreiben.